### PR TITLE
[java] New rule: AvoidReassigningLoopVariables

### PR DESCRIPTION
**Pull request by kris-scheibe**
This implements the new AvoidReassigningLoopVariables rule proposed in #1518
- the rule checks for and foreach loops
- "for" loops: configuration to completely allow/deny changing the variable in the loop body or allow only incrementing/decrementing the variable
- "foreach" loops: configuration to completely allow/deny changing the variable in the loop body or allow changing it in the first statement
- test-cases to verify the allow, deny, and intermediate configuration options

**Rationale for the intermediate configuration options:**

Incrementing or decrementing a "for" loop variable is fairly common, e.g. when skipping/re-reading elements. Setting the variable directly is either a bug or an indicator that the loop is overly complicated. This ensures that the loop variable in general runs through the range of values as given in the header.

Changing the value of the "foreach" loop variable is usually done to clean up the values in the collection. This could also be done by changing/copying the values into a new collection or using a stream().map().forEach() construct; doing it inline is simply more efficient. Changing the variable later in the loop body can be confusing as it isn't immediately visible which part of the loop uses the original value and which part uses the changed value.

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 129 new violations and 5 new errors,<br />
removes 0 violations and 2 errors. <a href="http://chunk.io/pmd/e78bc2a410d54d66831fa9b3e4b89452/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    private static final PropertyDescriptor<ForReassignOption> FOR_REASSIGN
+            = enumProperty("forReassign", FOR_REASSIGN_VALUES)
+            .defaultValue(ForReassignOption.DENY)
+            .desc("how/if for control variables may be reassigned")
+            .build();
+
+    public AvoidReassigningLoopVariablesRule() {
+        definePropertyDescriptor(FOREACH_REASSIGN);
+        definePropertyDescriptor(FOR_REASSIGN);
+    }
```
Comment (oowekyala) [inline pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java]: Since your Rule doesn't store state between visits of different AST nodes (in fact you're in the special case where you only visit one node type), you can use the rule chain to get a big performance boost. The only thing to do here is to add a call to `addRuleChainVisit(ASTLocalVariableDeclaration.class);` in your constructor

File: pmd-java/src/main/resources/category/java/bestpractices.xml [hunk 2]
```diff
@@ -167,6 +167,29 @@ class Foo {
         </example>
     </rule>
 
+    <rule name="AvoidReassigningLoopVariables"
+          since="6.11.0"
+          message="Avoid reassigning the loop control variable ''{0}''"
+          class="net.sourceforge.pmd.lang.java.rule.bestpractices.AvoidReassigningLoopVariablesRule"
+          externalInfoUrl="${pmd.website.baseurl}/pmd_rules_java_bestpractices.html#avoidreassigningloopvariables">
+        <description>
+Reassigning loop control variables can lead to hard-to-find bugs. Prevent or limit how these variables can be changed.
```
Comment (oowekyala) [inline pmd-java/src/main/resources/category/java/bestpractices.xml]: The possible values for your properties, how their affect the rule, and their defaults should be documented here. Possibly also add comments to your example code to show how different options would cause a different result.
We'll improve how properties are documented as part of #1432, but for now the rule description is the best way to do that

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+    <test-code>
+        <description>no violation: incremented 'for' loop variable</description>
+        <rule-property name="forReassign">skip</rule-property>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        for (int i=0; i < 10; i++) {
+            doSomethingWith(i);
+            i++;
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: I don't see a real reason for allowing to skip these. In this case and nearly all other test cases, the increment is executed at the end of *every* iteration, so it would be clearer to remove it and replace the for update with `i += 2`.

Wrapping the increment inside another control statement (like an `if`) should not trigger the rule though, e.g.
```java
        for (int i=0; i < 10; i++) {
           doSomethingWith(i);
           if (something()) i++; // skip one sometimes
        }
```
I'm not sure we can do that precisely before having some reliable local control flow analysis. E.g. the following should also not trigger the rule:
```java
        for (int i=0; i < 10; i++) {
           doSomethingWith(i);
           if (something()) continue; 
           i++;
        }
```
I think, we'd better keep it like it is until we can do better

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        <rule-property name="foreachReassign">firstOnly</rule-property>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>7</expected-linenumbers>
+        <code><![CDATA[
+public class Foo {
+    void foo(int bar) {
+        int[] ints = {1, 2, 3};
+        for (int i : ints) {
+            i++;
+            doSomethingWith(i);
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: Would `doSomethingWith(i++)` be flagged here? In fact more generally, would we detect assignments that are in an arbitrary expression, and not just as a simple statement? It seems to me this should be handled as well

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+            i++; // not OK
+        }
+    }
+}
+     ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>violation: pre-increment 'foreach' loop variable multiple times</description>
+        <rule-property name="foreachReassign">firstOnly</rule-property>
```
Comment (oowekyala) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: I'm not sure firstOnly is really useful for foreach loops. My 2 cents: 
* AFAIK it's uncommon to use a foreach loop to iterate over an array of numbers anyway
* The  `i` in foreach loops is not an index variable, but an item of an iterable. In any case I can think of, it's confusing and not useful to reassign it:
   * it doesn't change the value in the iterable
   * the same thing can be achieved by `doSomethingWith(i + 5)`, without reassignment.
   * if really needed, you can always declare another variable inside the loop `int shifted = i + 5`, which clarifies the code and preserves the original item

In fact, I'm interested in the behaviour of the rule with other data types, e.g.
```java
    void foo(List<Node> bar) {
        for (Node n : bar) {
            n = n.jjtGetParent(); // I can't think of a coding standard where this wouldn't be considered bad practice
            // better do Node parent = n.jjtGetParent();
            doSomethingWith(n);
        }
    }
```

Do you really see a use case for configurability of the reassignment policy in foreach loops, besides `allow`/`deny`?

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java (hunk 1 above)
Comment (kris-scheibe) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java]: Nice, I didn't know about that... Done!

File: pmd-java/src/main/resources/category/java/bestpractices.xml (hunk 2 above)
Comment (kris-scheibe) [reply pmd-java/src/main/resources/category/java/bestpractices.xml]: I documented the properties/their possible values, including a description/explanation for the intermediate options.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml (hunk 3 above)
Comment (kris-scheibe) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: I've looked into that and I think it's actually possible to do some limited kind of control flow analysis:

I can check if either
- the skip happens in some kind of control statement or 
- the skip is after a continue statement.

IMO, there are only a few cases where plain control flow analysis could work (e.g. if the skip is executed in every branch of an if/else or switch/case).
Apart from that, control statements are either constant conditions which can/should be covered by other rules or use variables or external logic and can't be analyzed statically.

I updated the tests accordingly. So far it seems  to be working!

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml (hunk 4 above)
Comment (kris-scheibe) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: With the new limited flow analysis `doSomethingWith(i++)` would be reported, unless it happens in a control statement. `doSomethingWith(i = 5)` would be reported, too.

In the current implementation, assignments in expressions aren't handled differently than simple statements.

I'm not sure if we should treat `if (something()) i++;` differently than `if (something()) doSomethingWith(i++)` -- at least in the scope of this rule.
Checking for assignments inside other expressions seems like a good candidate for a separate rule.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml (hunk 5 above)
Comment (kris-scheibe) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: OK, iterating over numbers really isn't a good example.

What I want to achieve with having `firstOnly` is the ability to write
```java
for (String stringWithSpaces : listOfStrings()) {
    final String string = stringWithSpaces.trim();
    // keep working with string
}
```
as
```java
for (String string : listOfStrings()) {
    string = string.trim();
    // keep working with string
}
```

In the first example, `stringWithSpaces` still exists and could be used (even if it shouldn't). In the second example the code will always use the cleaned value.

The `string.trim()` is only representative for some kind of transformation/clean-up and only useful if the variable is used more than once.
So this version
```java
    void foo(List<Node> bar) {
        for (Node n : bar) {
            doSomethingWith(n.jjtGetParent());
        }
    }
```
is preferable to the reassignment of the loop variable.

It is an edge-case, but a valid one, I think.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml [hunk 6]
```diff
@@ -0,0 +1,477 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<test-data
+        xmlns="http://pmd.sourceforge.net/rule-tests"
+        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        xsi:schemaLocation="http://pmd.sourceforge.net/rule-tests http://pmd.sourceforge.net/rule-tests_1_0_0.xsd">
+    <test-code>
+        <description>violation: reassigned 'for' loop variable</description>
+        <expected-problems>1</expected-problems>
+        <expected-linenumbers>5</expected-linenumbers>
+        <code><![CDATA[
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: To verify, that the message is produced correctly and the loop's variable name is used, you can add also an assert for the message:

```xml
<expected-messages>
  <message>Avoid reassigning the loop control variable 'i'</message>
</expected-messages>
```

I think, it's enough, if this is checked for once test case - we don't need to verify this for all.

Comment (kris-scheibe) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: Done!

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml (hunk 3 above)
Comment (kris-scheibe) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: I've changed the logic to make sure that changing the variable in the expression of a control statement is also reported. So ´if(foo(i++)) doSomething();´ triggers the rule, while ´if (something()) i++;´ does not.

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java [hunk 7]
```diff
@@ -0,0 +1,373 @@
...
+                // wrong type of operator
+                continue;
+            }
+
+            if (ignoreNode(operator, loopBody, ignoreFlags)) {
+                continue;
+            }
+
+            final ASTPrimaryExpression primaryExpression = operator.jjtGetParent().getFirstChildOfType(ASTPrimaryExpression.class);
+            final ASTName name = primaryExpression.getFirstDescendantOfType(ASTName.class);
```
[the remark this answers is not in the document]
Comment (kris-scheibe) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java]: Looking through the AST: By traversing through `ASTPrimaryExpression -> ASTPrimaryPrefix -> ASTName` I can make sure that I have variable's name (or null, if the expression is `this.` or `super.`)
Additionally checking for `ASTPrimaryExpression -> ASTPrimarySuffix` allows detecting arrays.
This method doesn't interfere with detecting `array[i = 4] =...`: It's treated the same way as `doSomethingWith(i = 4)`
.
I've added these checks and additional test cases.


But I'm a little bit confused by the representations of different kinds of field accesses in the AST.
Given this code:
```java
public class Foo {
    void foo() {
        for(String[] bar : bars()) {
            bar = new String[10]; // this is the only actual reassignment of bar
            bar[0] = "foo";
            bar.length = 5; // let's assume that array.length is writable...
            this.bar = 5;
        }
    }
}
```
The four assignments are parsed as:
- `PrimaryPrefix/Name[@Image='bar']`
- `PrimaryPrefix/Name[@Image='bar']` and `PrimarySuffix[@ArrayDereference=true]`
- `PrimaryPrefix/Name[@Image='bar.length']`
- `PrimaryPrefix[@ThisModifier=true]` and `PrimarySuffix[@Image='bar']`

Shouldn't the third one be parsed as `PrimaryPrefix/Name[@Image='bar']` and `PrimarySuffix[@Image='length']`?
I'm not sure if this is a bug or an intended feature/behavior...

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml (hunk 5 above)
Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: You've convinced me. I originally thought keeping the two variables in scope was an advantage, but you're right that for long loops it's better to reduce the number of variables in scope. Thanks for explaining

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml (hunk 3 above)
Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: Very well done. We can perfect that incrementally based on future false-positive/negative reports, but for now this is a good starting point

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml (hunk 4 above)
Comment (oowekyala) [reply pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/bestpractices/xml/AvoidReassigningLoopVariables.xml]: Hum yes, in fact we already have the rule [AssignmentInOperand](https://pmd.github.io/pmd-6.10.0/pmd_rules_java_errorprone.html#assignmentinoperand). Your rule does well to ignore `if (something()) doSomethingWith(i++)`, it will be flagged by AssignmentInOperand if the user chooses

File: pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (adangel) [reply pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/AvoidReassigningLoopVariablesRule.java]: Not sure, whether that's a bug or not. It probably just evolved in that way and now it is that way - this/super probably have been integrated later into the grammar. I've added a note to #1019 as changing this would break a lot of things... so for now we need to live with it, but we can change it for 7.0.0.
